testing-common/src/main/java/io/opentelemetry/instrumentation/testing/junit/http/AbstractHttpServerTest.java

```diff
@@ -0,0 +1,707 @@
+    TextMapSetter<HttpRequestBuilder> setter = HttpRequestBuilder::header;
+
+    for (int i = 0; i < count; i++) {
+      int index = i;
+      HttpRequestBuilder request =
+          HttpRequest.builder()
+              // Force HTTP/1 via h1c so upgrade requests don't show up as traces
+              .get(endpoint.resolvePath(address).toString().replace("http://", "h1c://"))
+              .queryParam(ServerEndpoint.ID_PARAMETER_NAME, index);
+
+      testing.runWithSpan(
+          "client " + index,
+          () -> {
+            Span.current().setAttribute(ServerEndpoint.ID_ATTRIBUTE_NAME, index);
+            propagator.inject(Context.current(), request, setter);
+            client.execute(request.build()).aggregate().thenRun(latch::countDown);
+          });
+    }
+    latch.await();
+
+    assertHighConcurrency(count);
+  }
+
+  protected void assertHighConcurrency(int count) {
+    ServerEndpoint endpoint = INDEXED_CHILD;
+    List<Consumer<TraceAssert>> assertions = new ArrayList<>();
+    for (int i = 0; i < count; i++) {
+      assertions.add(
+          trace -> {
+            SpanData rootSpan = trace.getSpan(0);
+            // Traces can be in arbitrary order, let us find out the request id of the current one
+            int requestId = Integer.parseInt(rootSpan.getName().substring("client ".length()));
+
+            List<Consumer<SpanDataAssert>> spanAssertions = new ArrayList<>();
+            spanAssertions.add(
+                span ->
+                    span.hasName(rootSpan.getName())
+                        .hasKind(SpanKind.INTERNAL)
+                        .hasNoParent()
+                        .hasAttributesSatisfying(
+                            attrs ->
+                                assertThat(attrs)
+                                    .containsEntry(ServerEndpoint.ID_ATTRIBUTE_NAME, requestId)));
+            spanAssertions.add(
+                span -> assertIndexedServerSpan(span, requestId).hasParent(rootSpan));
+
+            if (options.hasHandlerSpan.apply(endpoint)) {
+              spanAssertions.add(
+                  span -> assertHandlerSpan(span, "GET", endpoint).hasParent(trace.getSpan(1)));
+            }
+
+            int parentIndex = spanAssertions.size() - 1;
+            spanAssertions.add(
+                span ->
+                    assertIndexedControllerSpan(span, requestId)
+                        .hasParent(trace.getSpan(parentIndex)));
+
+            trace.hasSpansSatisfyingExactly(spanAssertions);
+          });
+    }
+    testing.waitAndAssertTraces(assertions);
+  }
+
+  protected void assertTheTraces(
+      int size,
+      String traceId,
+      String parentId,
+      String method,
+      ServerEndpoint endpoint,
+      String errorMessage,
+      AggregatedHttpResponse response) {
+    List<Consumer<TraceAssert>> assertions = new ArrayList<>();
+    for (int i = 0; i < size; i++) {
```

Comment:
This doesn't have all the span types that groovy version has.

Reply:
ah, this is why we still have the groovy implementations of those two methods? can you add a comment either here or there explaining, e.g. if there's some current limitation that makes this not possible, or if it's just a TODO to be implemented?

Reply:
The main issue is that I don't know how to completely unify java and groovy asserts. Currently groovy can call java asserts but not the other way around.
Here I just didn't add stuff that wasn't needed by the grails test. These can be added later when needed.